File: dotnet/test/common/MiscTest.cs
using NUnit.Framework;
using OpenQA.Selenium.Environment;
using System.Collections.Generic;

namespace OpenQA.Selenium
{
    [TestFixture]
    public class MiscTest : DriverTestFixture
    {
        [Test]
        public void ShouldReturnTitleOfPageIfSet()
        {
            driver.Url = xhtmlTestPage;
            Assert.That(driver.Title, Is.EqualTo("XHTML Test Page"));

            driver.Url = simpleTestPage;
            Assert.That(driver.Title, Is.EqualTo("Hello WebDriver"));
        }

        [Test]
        public void ShouldReportTheCurrentUrlCorrectly()
        {
            driver.Url = macbethPage;
            Assert.AreEqual(macbethPage, driver.Url);

            driver.Url = simpleTestPage;
            Assert.AreEqual(simpleTestPage, driver.Url);

            driver.Url = javascriptPage;
            Assert.AreEqual(javascriptPage, driver.Url);
        }

        [Test]
        public void ShouldReturnTagName()
        {
            driver.Url = formsPage;
            IWebElement selectBox = driver.FindElement(By.Id("cheese"));
            Assert.That(selectBox.TagName.ToLower(), Is.EqualTo("input"));
        }

        [Test]
        public void ShouldReturnTheSourceOfAPage()
        {
            string pageSource;
            driver.Url = simpleTestPage;
            pageSource = driver.PageSource.ToLower();

            Assert.That(pageSource, Does.StartWith("<html"));
            Assert.That(pageSource, Does.EndWith("</html>"));
            Assert.That(pageSource, Does.Contain("an inline element"));
            Assert.That(pageSource, Does.Contain("<p id="));
            Assert.That(pageSource, Does.Contain("lotsofspaces"));
            Assert.That(pageSource, Does.Contain("with document.write and with document.write again"));
        }

        [Test]
        [IgnoreBrowser(Browser.Chrome, "returns XML content formatted for display as HTML document")]
        [IgnoreBrowser(Browser.Edge, "returns XML content formatted for display as HTML document")]
        [IgnoreBrowser(Browser.Safari, "returns XML content formatted for display as HTML document")]
        [IgnoreBrowser(Browser.IE, "returns XML content formatted for display as HTML document")]
        [IgnoreBrowser(Browser.EdgeLegacy, "returns XML content formatted for display as HTML document")]
        [IgnoreBrowser(Browser.Opera)]
        public void ShouldBeAbleToGetTheSourceOfAnXmlDocument()
        {
            driver.Url = simpleXmlDocument;
            string source = driver.PageSource.ToLower();
            source = System.Text.RegularExpressions.Regex.Replace(source, "\\s", string.Empty);
            Assert.AreEqual("<xml><foo><bar>baz</bar></foo></xml>", source);
        }

        // Test is ignored for all browsers, but is kept here in the source code for
        // ease of comparison to Java test suite.
        //[Test]
        //[IgnoreBrowser(Browser.All, "issue 2282")]
        //public void StimulatesStrangeOnloadInteractionInFirefox()
        //{
        //    driver.Url = documentWrite;

        //    // If this command succeeds, then all is well.
        //    driver.FindElement(By.XPath("//body"));

        //    driver.Url = simpleTestPage;
        //    driver.FindElement(By.Id("links"));
        //}

        [Test]
        public void ClickingShouldNotTrampleWOrHInGlobalScope()
        {
            driver.Url = EnvironmentManager.Instance.UrlBuilder.WhereIs("globalscope.html");
            List<string> values = new List<string>() { "w", "h" };

            foreach (string val in values)
            {
                Assert.AreEqual(val, GetGlobalVar(driver, val));
            }

            driver.FindElement(By.Id("toclick")).Click();

            foreach (string val in values)
            {
                Assert.AreEqual(val, GetGlobalVar(driver, val));
            }
        }

        private string GetGlobalVar(IWebDriver driver, string value)
        {
            object val = ((IJavaScriptExecutor)driver).ExecuteScript("return window." + value + ";");
            return val == null ? "null" : val.ToString();
        }
    }
}

File: java/test/org/openqa/selenium/MiscTest.java
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.openqa.selenium;

import static org.assertj.core.api.Assertions.assertThat;
import static org.openqa.selenium.testing.drivers.Browser.ALL;
import static org.openqa.selenium.testing.drivers.Browser.CHROME;
import static org.openqa.selenium.testing.drivers.Browser.EDGE;
import static org.openqa.selenium.testing.drivers.Browser.IE;
import static org.openqa.selenium.testing.drivers.Browser.SAFARI;

import org.junit.jupiter.api.Test;
import org.openqa.selenium.testing.Ignore;
import org.openqa.selenium.testing.JupiterTestBase;
import org.openqa.selenium.testing.NotYetImplemented;

class MiscTest extends JupiterTestBase {

  @Test
  void testShouldReturnTitleOfPageIfSet() {
    driver.get(pages.xhtmlTestPage);
    assertThat(driver.getTitle()).isEqualTo(("XHTML Test Page"));

    driver.get(pages.simpleTestPage);
    assertThat(driver.getTitle()).isEqualTo("Hello WebDriver");
  }

  @Test
  void testShouldReportTheCurrentUrlCorrectly() {
    driver.get(pages.simpleTestPage);
    assertThat(driver.getCurrentUrl()).isEqualToIgnoringCase(pages.simpleTestPage);

    driver.get(pages.javascriptPage);
    assertThat(driver.getCurrentUrl()).isEqualToIgnoringCase(pages.javascriptPage);
  }

  @Test
  void shouldReturnTagName() {
    driver.get(pages.formPage);
    WebElement selectBox = driver.findElement(By.id("cheese"));
    assertThat(selectBox.getTagName()).isEqualToIgnoringCase("input");
  }

  @Test
  void testShouldReturnTheSourceOfAPage() {
    driver.get(pages.simpleTestPage);

    String source = driver.getPageSource().toLowerCase();

    assertThat(source)
        .contains(
            "<html",
            "</html",
            "an inline element",
            "<p id=",
            "lotsofspaces",
            "with document.write and with document.write again");
  }

  @Test
  @Ignore(value = CHROME, reason = "returns XML content formatted for display as HTML document")
  @Ignore(value = EDGE, reason = "returns XML content formatted for display as HTML document")
  @NotYetImplemented(
      value = SAFARI,
      reason = "returns XML content formatted for display as HTML document")
  @Ignore(IE)
  public void testShouldBeAbleToGetTheSourceOfAnXmlDocument() {
    driver.get(pages.simpleXmlDocument);
    String source = driver.getPageSource().toLowerCase();
    assertThat(source).isEqualToIgnoringWhitespace("<xml><foo><bar>baz</bar></foo></xml>");
  }

  @Test
  @Ignore(value = ALL, reason = "issue 2282")
  public void testStimulatesStrangeOnloadInteractionInFirefox() {
    driver.get(pages.documentWrite);

    // If this command succeeds, then all is well.
    driver.findElement(By.xpath("//body"));

    driver.get(pages.simpleTestPage);
    driver.findElement(By.id("links"));
  }

  @Test
  void testClickingShouldNotTrampleWOrHInGlobalScope() {
    driver.get(appServer.whereIs("globalscope.html"));
    String[] vars = new String[] {"w", "h"};
    for (String var : vars) {
      assertThat(getGlobalVar(driver, var)).isEqualTo(var);
    }
    driver.findElement(By.id("toclick")).click();
    for (String var : vars) {
      assertThat(getGlobalVar(driver, var)).isEqualTo(var);
    }
  }

  private String getGlobalVar(WebDriver driver, String var) {
    Object val = ((JavascriptExecutor) driver).executeScript("return window." + var + ";");
    return val == null ? "null" : val.toString();
  }
}
